### ISSUE TITLE
Kerollmops improvement list

### ISSUE THREAD

**Issue by c0gent** [cogciprocate/ocl#89]
[https://gist.github.com/Kerollmops/68fc37f049c57efc7c77e420da300e3b](https://gist.github.com/Kerollmops/68fc37f049c57efc7c77e420da300e3b)

From: https://github.com/cogciprocate/ocl/issues/85#issuecomment-355814751

**Comment by c0gent** [comment]
This feedback is very valuable to me. I don't get enough and I really appreciate this kind of constructive criticism.

- Global:
	- Everything need to work with the Future system
            - *I think you mean documentation. Added to TODO.*
	- Remove all panics (e.g. Platform)
            - *Panics are a useful and important escape hatch for unrecoverable problems: https://doc.rust-lang.org/book/second-edition/ch09-03-to-panic-or-not-to-panic.html. These are programmer errors that cannot be recovered from at runtime.*
	- Remove all useless `Display` implementations
            - *Agreed. Added to TODO.*
	- Make every builders become mutable builders
            - *In many cases this is impractical, impossible, or unnecessary. Builders come in several valid styles: https://aturon.github.io/ownership/builders.html. Please explain better what you mean.*
	- Reduce the visible types at the entrypoint of the doc
            - *Added to TODO.*
	- Create an Index type/Wrapper
            - *Please clarify.*
	- Remove useless lifetimes
            - *Please provide example.*
	- Add examples usage to a maximum of methods
            - *Agreed, more example code is needed throughout the project. Added to TODO.*

- Documentation:
	- Add a comment to the `Cargo.toml` file that reminds to bump the semvers of `html_root_url`
            - *Good idea. Added to TODO.*
	- Don't use short/reduced/cutted/unclear names for modules or types (c.g. prm for primitives)
            - *Why not? Often a shortened name is perfectly adequate.*

- Buffer:
	- Remove the `new` method
            - *Why? The new method is perfectly useful. Do you think it should be private? It may be useful for someone to create a buffer directly rather than use the builder.*
	- Why does it have a `default_queue` ?
            - *`::default_queue` is used to return a reference to the currently set default queue. It's useful for a variety of purposes, such as ensuring  multiple tasks use the same queue without have to store the queue separately.*
	- `create_sub_buffer` don't use `SpatialDims` for `origin`
            - *That's because using a multi-dimensional offset is ambiguous and conflicts with the way offsets into rectangular and cubic buffers works. I agree that `SpatialDims` needs to be addressed.*
	- `create_sub_buffer` reword the documentation
            - *The documentation was adapted from the official documentation. It could probably use some work. Added to TODO.*
	- Explain why a `default_queue` can be specified and why it's optional
            - *Default queues could be explained better. Added to TODO.*
	- Remove `from_gl_buffer` or change the arguments
            - *Why? This is a useful method. What's wrong with the arguments? Please clarify.*
	- Change the way `read`, `write` methods works, use futures ?
            - *Please clarify.*

- BufferBuilder:
	- Refacto the `context` along with the `queue` method
            - *Refactor? Please clarify.*
	- Refacto and clarify the `host_data` method
            - *Please clarify.*
	- Remove `dims` and/or use another type to represent the data with 1/2/3 dimensions
            - *`SpatialDims` needs a redesign or replacement. Added to TODO.*
	- Unify `fill_val` and `fill_event` because that's the same thing, use futures
            - *They are not the same thing and have been separated for good reason. I'll will add more documentation to clarify. Added to TODO.*

- BufferCmd:
	- Remove the `new` method
            - *I'll make it private. Added to TODO*
	- Remove useless panics when the command operation has already been defined
            - *See: 'Remove all panics' above.*
	- Remove `read_async`
            - *Added to TODO.*
	- Add more informations on the `map` method
            - *Added to TODO.*
	- Change the way `gl_aquire` and `gl_release` works (e.g. Remove antoher type, more clear)
            - *I honestly have never used either function. I'll look into this. Added to TODO.*
	- Why does `block` is unsafe ?
            - *Because setting `block` to false is potentially unsafe due to asynchronous timings.*
	- Refacto the `rect` method and doesn't panic
            - *Please clarify.*
	- Remove `ewait_opt`
            - *Added to TODO.*
	- Add more explanations to the `ewait`/`enew` methods
            - *Added to TODO.*

- Mem/MapFlags:
	- Follow the `std::OpenOptions` for `ocl::Mem/MapFlags` (https://doc.rust-lang.org/std/fs/struct.OpenOptions.html)
            - *Please explain. Mem/MapFlags already follow the same pattern (e.g. `MemFlags::new().read_only().host_write_only()`)*

- Device:
	- `Device::first` will panic and don't needs to
            - *Added to TODO.*

- Platform:
	- Don't use `Vec` but `&[]` in `list_from_core`
            - *Compiler will reuse allocation the way it is. Will make pub(crate) instead. Added to TODO*
	- Don't return raw `String` from `profile` and `extensions`
            - *`PlatformInfo::Profile` and `PlatformInfo::Extensions` queries returned owned strings.*

- Kernel:
	- `new` use `Into<&str>` instead of the useless `Into<String>`
            - *`Into<String>` is far from useless. It allows the caller to pass either an owned string or a reference, avoiding an allocation in the former case.*
	- Try to make a better API to arg passing with real typed arguments
            - *I agree that the API needs work/redesign. I've always wanted to change it but have never been able to decide on a better design. Any suggestions are very welcome. I'm happy to spend more time discussing this. Added to TODO.*
	- Remove unsafe function from top level function
            - *Please clarify.*
	- Change `gwo` into `set_gwo` and `get_gwo` into `gwo`
            - *Redesign needed. See prior note. Added to TODO.*
	- Rename `core` into `as_core`
            - *All `::core` functions need renaming to `::as_core`. Added to TODO.*
	- Rename `by_idx` into `by_index` for more clarity
            - *Agreed. Added to TODO.*

- RwVec:
	- Add an `is_empty` method to the type
            - *Added to TODO.*

- EventArray:
	- Remove the `push_some`
            - *Added to TODO.*

**Comment by c0gent** [comment]
Just pinging you, @Kerollmops :)

**Comment by Kerollmops** [comment]
I'm actually really busy but I added a reminder to add more in this list. I will come back this week probably.  😃

**Comment by Kerollmops** [comment]
This is very complex to make a clear answer. Their is too many points of conversations and I am not done.
I want to find something like a Trello or one issue by improvement but in another repo to limit the noise.

**Comment by c0gent** [comment]
Yeah, I was thinking about making a [Github Project](https://github.com/cogciprocate/ocl/projects) for it. Trello would be great too.

**Comment by Kerollmops** [comment]
And why not RFCs ?
Do you think it is overkill or soemthing ?

**Comment by c0gent** [comment]
RFC would be fine, but yes, a bit overkill. In this case I think you should create a Trello board where the two of us can sort out the majority of the small issues. If any large questions remain, we can create an issue in this repo, or create an RFC repo.

**Comment by c0gent** [comment]
Contact me via: https://gitter.im/cogciprocate/ocl if you need my attention. I usually keep the chat open so it's faster than communicating via github or e-mail :)

**Comment by Kerollmops** [comment]
We are know talking on [this Trello board](https://trello.com/b/kflyz03A/possible-improvements).

**Comment by Zolmeister** [comment]
Implement `std::error::Error` is for `ocl::Error`
